### ISSUE TITLE
5.1 and 7.1 surround sound not working

### ISSUE THREAD

**Issue by lexum0** [godotengine/godot#9494]
**Operating system or device - Godot version:**
Windows 10 - Godot 2 & 3

**Issue description:**
Trying to play spatial sound only works as left and right sound (stereo) ignoring the surround setup. It seems that in the code it's hardcoded to STEREO and the needed code is commented or has FIXME comments

Also there's no info on how to use surround sound, so maybe it's just not being set up correctly. I changed the audio Project Settings to 'surround' instead of 'stereo' with no luck

**Comment by akien-mga** [comment]
As far as I know 5.1 and 7.1 surround sound is not implemented in 2.1, and it likely won't be. It's supposed to be implemented in the master branch though, but @reduz did not have the required hardware to try it, so there might be some issues to fix.

**Comment by reduz** [comment]
there is support in 3.0 but i don have any hardware to test, and the
drivers need to be modified.

**Comment by lexum0** [comment]
Tell me how and I can do the testing

**Comment by marcelofg55** [comment]
@lexum0 Hey, I've been working on some surround fixes (#10568), could you help me test it?

**Comment by lexum0** [comment]
Sure! What can I do?

**Comment by marcelofg55** [comment]
You can test the new changes with this build: https://www.dropbox.com/s/ubgdstavst7kni9/godot.windows.tools.32.exe?dl=0, or if you prefer to compile the git branch with the changes is: https://github.com/marcelofg55/godot/tree/surround_fixes
Let me know if you find any issue with it, thanks!

**Comment by lexum0** [comment]
I'm trying to run it but it keeps giving me this error:
ERROR: bind_methodfi: Class MeshLibrary already has a method remove_item
   At: core/class_db.cpp:1182

and running the game crashes and won't run. 
I compiled it for macOS Sierra. I'll give it a try on windows

**Comment by marcelofg55** [comment]
What's your surround speaker configuration on windows? 5.1 or 7.1?

**Comment by lexum0** [comment]
on Windows 7.1, on Mac 5.1

**Comment by marcelofg55** [comment]
Could you try this patch on top of the surround_fixes branch? [channels.patch.zip](https://github.com/godotengine/godot/files/1278859/channels.patch.zip)
It will show how much channels godot is detecting

**Comment by lexum0** [comment]
it's detecting 6 or 8 channels, depending on how I set up the Mac. But no sound gets out, I think it's due to the error

**Comment by marcelofg55** [comment]
And on windows?

**Comment by lexum0** [comment]
Just tested on Windows with 5.1, still stereo sound. I don't have windows set up for compiling, so I can't test the patch on windows

**Comment by lexum0** [comment]
Is there a way to visualize or read what each channel is supposed to output? Something like the db meter. If I can read a number for each channel, that would be super useful for testing, also for knowing what each speaker is supposed to be playing

**Comment by reduz** [comment]
@marcelofg55 how is this issue doing?

**Comment by marcelofg55** [comment]
Surround is working now, so this should be closed I think. There is still some more work left to do with it: detect user speaker channel mappings and check/fix the 3d audio position logic. I'll keep working on it sometime.

**Comment by zwostein** [comment]
It seems I just came across the same issue:

When using AudioStreamPlayer3D in a 5.1 surround configuration the sound does not seem to be positioned correctly. The panning between left and right side is abrupt and it is hard to hear from where the sound actually comes from. It is fine when I am switching to a stereo configuration, though.

However, I was playing around a little and implemented a different multichannel panning algorithm. Maybe someone with a 5.1 or 7.1 setup wants to give PR https://github.com/godotengine/godot/pull/30234 a try?

On Linux using Pulseaudio I also had to explicitly set the channel map - see PR https://github.com/godotengine/godot/pull/30232 .
